dipy/denoise/denspeed.pyx

```diff
@@ -153,49 +157,70 @@ cdef double process_block(double [:, :, ::1] arr,
+        cnp.npy_intp PS = P * 2 + 1
+        cnp.npy_intp TS = PS + BS - 1
 
     cnt = 0
     sumw = 0
     patch_vol_size = (P + P + 1) * (P + P + 1) * (P + P + 1)
+    block_vol_size = (B + B + 1) * (B + B + 1) * (B + B + 1)
 
-    W = <double *> malloc(BS * BS * BS * sizeof(double))
-    cache = <double *> malloc(BS * BS * BS * sizeof(double))
-    sigma_block = <double *> malloc(BS * BS * BS * sizeof(double))
+    W = <double *> malloc(PS * PS * PS * sizeof(double))
+    cache = <double *> malloc(TS * TS * TS * sizeof(double))
+    sigma_block = <double *> malloc(TS * TS * TS * sizeof(double))
 
     # (i, j, k) coordinates are the center of the static patch
     # copy block in cache
-    copy_block_3d(cache, BS, BS, BS, arr, i - B, j - B, k - B)
-    copy_block_3d(sigma_block, BS, BS, BS, sigma, i - B, j - B, k - B)
+    copy_block_3d(
+        cache,
+        PS + BS - 1,
+        PS + BS - 1,
+        PS + BS - 1,
+        arr,
+        i - P - B,
+        j - P - B,
+        k - P - B)
+    copy_block_3d(
+        sigma_block,
+        PS + BS - 1,
+        PS + BS - 1,
+        PS + BS - 1,
+        arr,
+        i - P - B,
+        j - P - B,
+        k - P - B)
 
     # calculate weights between the central patch and the moving patch in block
     # (m, n, o) coordinates are the center of the moving patch
     # (a, b, c) run inside both patches
-    for m in range(P, BS - P):
-        for n in range(P, BS - P):
-            for o in range(P, BS - P):
+    for m in range(-P, P + 1):
+        for n in range(-P, P + 1):
+            for o in range(-P, P + 1):
 
                 summ = 0
                 sigm = 0
 
                 # calculate square distance
-                for a in range(-P, P + 1):
-                    for b in range(-P, P + 1):
-                        for c in range(-P, P + 1):
+                for a in range(-B, B + 1):
+                    for b in range(-B, B + 1):
+                        for c in range(-B, B + 1):
 
                             # this line takes most of the time! mem access
-                            d = cache[(B + a) * BS * BS + (B + b) * BS + (B + c)] - cache[(m + a) * BS * BS + (n + b) * BS + (o + c)]
+                            d = cache[(P + B + a) * TS * TS + (P + B + b) * TS + (P + B + c)] - cache[
+                                (P + B + m + a) * TS * TS + (P + B + n + b) * TS + (P + B + o + c)]
                             summ += d * d
-                            sigm += sigma_block[(m + a) * BS * BS + (n + b) * BS + (o + c)]
+                            sigm += sigma_block[(P + B + m + a) * TS *
+                                                TS + (P + B + n + b) * TS + (P + B + o + c)]
 
-                denom = sqrt(2) * (sigm / patch_vol_size)**2
-                w = exp(-(summ / patch_vol_size) / denom)
+                denom = sqrt(2) * (sigm / block_vol_size)**2
+                w = exp(-(summ / block_vol_size) / denom)
                 sumw += w
```

Comment:
@Garyfallidis these will be changed to
`denom = sigm / block_vol_size`
`w = exp(-sqrt(summ/ block_vol_size)/denom)`